fil_fungible_token/src/token/mod.rs

```diff
@@ -278,8 +285,380 @@ where
+            .unwrap();
+        // burn the approved amount
+        token.burn(burner, treasury, burn_amount.clone()).unwrap();
+
+        // total supply decreased
+        let total_supply = token.total_supply();
+        assert_eq!(total_supply, TokenAmount::from(400_000));
+        // treasury balance decreased
+        let balance = token.balance_of(treasury).unwrap();
+        assert_eq!(balance, TokenAmount::from(400_000));
+        // burner approval decreased
+        let new_allowance = token.allowance(treasury, burner).unwrap();
+        assert_eq!(new_allowance, TokenAmount::zero());
+
+        // disallows another delegated burn as approval is zero
+        // burn the approved amount
+        token
+            .burn(burner, treasury, burn_amount)
+            .expect_err("unable to burn more than allowance");
+
+        // balances didn't change
+        let total_supply = token.total_supply();
+        assert_eq!(total_supply, TokenAmount::from(400_000));
+        let balance = token.balance_of(treasury).unwrap();
+        assert_eq!(balance, TokenAmount::from(400_000));
+        let new_allowance = token.allowance(treasury, burner).unwrap();
+        assert_eq!(new_allowance, TokenAmount::zero());
+    }
+
+    #[test]
+    fn it_cannot_burn_below_zero() {
+        let state_cid = Cid::default();
+        let mut token = Token::new(SharedMemoryBlockstore::new(), state_cid);
+        token.init_state().unwrap();
+
+        let treasury = 1;
+        let mint_amount = TokenAmount::from(1_000_000);
+        let burn_amount = TokenAmount::from(2_000_000);
+        token.mint(treasury, mint_amount).unwrap();
+        token.burn(treasury, treasury, burn_amount).unwrap_err();
+
+        // total supply remained same
+        let total_supply = token.total_supply();
+        assert_eq!(total_supply, TokenAmount::from(1_000_000));
+
+        // treasury balance remained same
+        let balance = token.balance_of(treasury).unwrap();
+        assert_eq!(balance, TokenAmount::from(1_000_000));
+    }
+
+    #[test]
+    fn it_allows_transfer() {
+        let state_cid = Cid::default();
+        let mut token = Token::new(SharedMemoryBlockstore::new(), state_cid);
+        token.init_state().unwrap();
+
+        let owner = 1;
+        let receiver = 2;
+        // mint 100 for owner
+        token.mint(owner, TokenAmount::from(100)).unwrap();
+        // TODO: token needs some injectable layer to handle and mock the receive hook behaviour
+        // transfer 60 from owner -> receiver
+        token
+            .transfer(owner, owner, receiver, TokenAmount::from(60))
+            .unwrap();
+
+        // owner has 100 - 60 = 40
+        let balance = token.balance_of(owner).unwrap();
+        assert_eq!(balance, TokenAmount::from(40));
+
+        // receiver has 0 + 60 = 60
+        let balance = token.balance_of(receiver).unwrap();
+        assert_eq!(balance, TokenAmount::from(60));
```

Comment:
- Test transfer of zero
- Attempt transfer negative
- Transfer to self with zero, and non-zero amounts
- Check the total supply is unchanged in all these cases